### PR TITLE
feat(identity): make rand optional

### PR DESCRIPTION
**Pull request by monoid**
## Description

Some restricted environments lack proper random source, and `getrandom` crate fails to compile.  However, random sources is needed only for random key generation which is not always required.

This change introduces `rand` feature flag that is gate to all the functionality that requires `rand` crate.

## Notes & open questions

The idea of this pull request was born when I tried to use the crate in NEAR contract.

## Change checklist

- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] A changelog entry has been made in the appropriate crates

### CODE REVIEW

**Comment by thomaseizinger** [comment]
Thanks! This is unfortunately a breaking change so I'd like to hold off on it for a while. We've only just released a breaking change to `libp2p-identity` 2 months ago.

Is this blocking you on something? Out of interest, what is your usecase where you don't have a randomness source? Do `libp2p-tls` or `libp2p-noise` even work without randomness?

**Comment by monoid** [comment]
> Thanks! This is unfortunately a breaking change so I'd like to hold off on it for a while. We've only just released a breaking change to `libp2p-identity` 2 months ago.
> 
> Is this blocking you on something?

Well, we can live with git branch deps for some time.  I've submitted it mostly because it can be useful for others.

> Out of interest, what is your usecase where you don't have a randomness source? Do `libp2p-tls` or `libp2p-noise` even work without randomness?

We have bunch of (normal) peers who do some coordinated task and submit their result on a chain where a smartcontract validates their outputs.  The smartcontract uses `libp2p-identity` to match public keys and peer IDs, not using any other libp2p library.

**Comment by thomaseizinger** [comment]
> > Out of interest, what is your usecase where you don't have a randomness source? Do `libp2p-tls` or `libp2p-noise` even work without randomness?
> 
> We have bunch of (normal) peers who do some coordinated task and submit their result on a chain where a smartcontract validates their outputs.  The smartcontract uses `libp2p-identity` to match public keys and peer IDs, not using any other libp2p library.

Thanks for sharing!

Not sure if that is viable but perhaps other libraries also need randomness and the target should register a randomness source? 

See https://docs.rs/getrandom/latest/getrandom/index.html#custom-implementations.

In any case we can leave this open until we are making another breaking change. `libp2p-identity` is the leaf dependency of the entire ecosystem of libp2p because `multiaddr` depends on it and breaking changes thus have quite the blast radius.

**Comment by thomaseizinger** [comment]
Note for myself: Create a separate milestone and link it to the next `rust-multiaddr` milestone.

File: identity/src/keypair.rs
Comment (thomaseizinger) [inline identity/src/keypair.rs]: Do we need to do this? We always build docs with all features enabled.

**Comment by thomaseizinger** [comment]
I am tempted to release this as a patch-release. Whilst technically a breaking change, it will only hit users that actively say `default-features = false` or when they depend on a crate that set `default-features = false`. Because features are additive, this can easily be resolved by just adding a dependency on `libp2p-identity` and activating the `rand` feature.

@mxinden @jxs What do you think of this idea?

File: identity/Cargo.toml
Comment (thomaseizinger) [inline identity/Cargo.toml]: I think I'd actually prefer to not have a `default` feature and instead automatically turn on the `rand` feature via `libp2p/Cargo.toml`.

File: identity/src/keypair.rs
Comment (monoid) [reply identity/src/keypair.rs]: `cargo test` runs it as a doctest, and fails when rand feature is not enabled. It can be set to `ignore` statically to make it simpler, but it would mean that this function is not tested anymore (cannot check now if it is tested elsewhere).

Comment (thomaseizinger) [reply identity/src/keypair.rs]: We always test with all features so it should still be tested, right?

File: identity/Cargo.toml
Comment (monoid) [reply identity/Cargo.toml]: Done.

File: identity/src/keypair.rs
Comment (monoid) [reply identity/src/keypair.rs]: Well, it is not very convenient if one works with particular crate and wants to test it with particular flag combination.  I've reverted the comment to the original version, however, I hope it will work for you.

Comment (thomaseizinger) [reply identity/src/keypair.rs]: Thank you! I agree that it isn't super convenient for that case but I'd rather not have complexity like this in the docs.

**Comment by thomaseizinger** [comment]
Can you add the `rand` feature here please?

https://github.com/libp2p/rust-libp2p/blob/d862b403834b3067606ced65b6cac47fffce705b/libp2p/Cargo.toml#L111

**Comment by monoid** [comment]
@thomaseizinger Done!

**Comment by jxs** [comment]
> I am tempted to release this as a patch-release. Whilst technically a breaking change, it will only hit users that actively say `default-features = false` or when they depend on a crate that set `default-features = false`. Because features are additive, this can easily be resolved by just adding a dependency on `libp2p-identity` and activating the `rand` feature.
> 
> @mxinden @jxs What do you think of this idea?

yeah makes sense to me Thomas, did sorta similar [with `refinery`](https://github.com/rust-db/refinery/pull/286#issuecomment-1719290664)

**Comment by thomaseizinger** [comment]
Thanks for fixing the tests @monoid !

As a last thing, can you bump the patch-version of `libp2p-identity` and write a short changelog entry? I'll merge and release this right away after.

**Comment by monoid** [comment]
@thomaseizinger done!

File: identity/Cargo.toml
Comment (mxinden) [inline identity/Cargo.toml]: I am confused. Why keep the `dep:rand` here? In case it is required for ed25519, why do we need the feature flag `#[cfg(feature = "rand")]` in `ed25519.rs`?

**Review by mxinden — COMMENTED** [review]
I am in favor of this patch.

> I am tempted to release this as a patch-release. Whilst technically a breaking change, it will only hit users that actively say `default-features = false` or when they depend on a crate that set `default-features = false`. Because features are additive, this can easily be resolved by just adding a dependency on `libp2p-identity` and activating the `rand` feature.

Say that we release this as `v0.2.5` today. Dependents e.g. https://github.com/sigp/lighthouse that use Dependabot will have a pull request pop up that upgrades their `libp2p-identity` version to `v0.2.5`. None of their dependencies set the `rand` feature on `libp2p-identity`. Thus their pull request will not compile. Am I missing something?

File: identity/Cargo.toml
Comment (monoid) [reply identity/Cargo.toml]: I've overlooked it on merge, thanks.
Actually, it would be great to have a test for it that just tries to build this crate to `wasm32-unknown-unknown`, and if `rand` will be used somehow, it will fail.

File: identity/src/ed25519.rs
Comment (mxinden) [inline identity/src/ed25519.rs:38]: More specifically I am talking about this line. The `rand` feature is `libp2p-identity` specific, correct? Dependents upgrading to `libp2p-identity` `v0.2.5` will not have the `libp2p-identity` `rand` feature enabled no?

**Comment by thomaseizinger** [comment]
> > I am tempted to release this as a patch-release. Whilst technically a breaking change, it will only hit users that actively say `default-features = false` or when they depend on a crate that set `default-features = false`. Because features are additive, this can easily be resolved by just adding a dependency on `libp2p-identity` and activating the `rand` feature.
> 
> Say that we release this as `v0.2.5` today. Dependents e.g. [sigp/lighthouse](https://github.com/sigp/lighthouse) that use Dependabot will have a pull request pop up that upgrades their `libp2p-identity` version to `v0.2.5`. None of their dependencies set the `rand` feature on `libp2p-identity`. Thus their pull request will not compile. Am I missing something?

Correct! We should pair it with a patch release of `libp2p` that enables the `rand` feature.

File: identity/CHANGELOG.md
Comment (thomaseizinger) [inline identity/CHANGELOG.md]:
```suggestion
- Add `rand` feature to gate methods requiring a random number generator, enabling use in restricted environments (e.g. smartcontracts).
  This feature is not enabled by default.
  See [PR 4349](https://github.com/libp2p/rust-libp2p/pull/4349).
```

File: identity/Cargo.toml
Comment (thomaseizinger) [inline identity/Cargo.toml]: We use workspace inheritance in the root `Cargo.toml`, can you also bump the referenced version there please?

File: identity/Cargo.toml
Comment (thomaseizinger) [inline identity/Cargo.toml]:
```suggestion
rsa = [ "dep:ring", "dep:asn1_der", "dep:rand", "dep:zeroize" ] # It is ok for RSA to depend on rand as it doesn't compile to WASM and similar environments anyway.
```

This allows easier sorting of these lines alphabetically.

File: identity/CHANGELOG.md
Comment (thomaseizinger) [reply identity/CHANGELOG.md]: It seems I cannot push to your branch so I can't resolve these comments myself unfortunately.

**Review by thomaseizinger — COMMENTED** [review]
I think you'll need to add the rand feature to `libp2p-identity in the `libp2p-swarm-test` crate. Also, can you please make a patch-bump in the `libp2p` crate? Otherwise there will be a fair few CI failures for our users because they will only get a dependabot update for `libp2p-identity`.

**Comment by thomaseizinger** [comment]
@monoid Looks like a few commits got re-written and merged into this PR which is why they are all showing up now. I'd recommend to do an interactive rebase and drop those, then force-push and merge a latest `origin/master` into your feature branch.

**Comment by thomaseizinger** [comment]
> @monoid Looks like a few commits got re-written and merged into this PR which is why they are all showing up now. I'd recommend to do an interactive rebase and drop those, then force-push and merge a latest `origin/master` into your feature branch.

Let me know if you need help! Happy to do it as well.

**Comment by monoid** [comment]
Oops, will try to clean it up.

**Comment by monoid** [comment]
@thomaseizinger seems to be fixed.

**Comment by thomaseizinger** [comment]
See https://github.com/fluencelabs/rust-libp2p/pull/54.

File: identity/Cargo.toml
Comment (thomaseizinger) [inline identity/Cargo.toml:38]:
```suggestion
rsa = ["dep:ring", "dep:asn1_der", "dep:rand", "dep:zeroize"]
```

It does depend on `rand` though, so I don't think we need this comment :)

**Comment by monoid** [comment]
Merged and merged again.

**Comment by thomaseizinger** [comment]
> Merged and merged again.

Awesome, thanks! Sorry for the hassle!

**Comment by monoid** [comment]
@thomaseizinger no problem!  This is a life of an actively developed project. :)

**Comment by thomaseizinger** [comment]
I think some unit tests still implicitly depend on the `rand` feature and are now failing because it is not there. I think you'll have to add a dedicated `dev-dependencies` entry for those where we activate the `rand` feature.

**Comment by monoid** [comment]
@thomaseizinger tests should work now.

**Review by mxinden — COMMENTED** [review]
Feel free to go ahead here @thomaseizinger. Thank you for driving this to both of you.

@thomaseizinger when you cut the patch release, would you mind keeping an eye open the repositories of our largest users to make sure they don't face any difficulties with the upgrade?

Projects that come to my mind are:

- https://github.com/paritytech/polkadot-sdk
  - Probably best to simply propose a patch on https://github.com/paritytech/polkadot-sdk/pull/1631/
- https://github.com/sigp/lighthouse
- https://github.com/ChainSafe/forest

**Comment by thomaseizinger** [comment]
> @thomaseizinger tests should work now.

Still a few more failures unfortunately. You can ignore the `semver` job, that one fails because this is a bit of a special situation with `libp2p-identity`.

**Comment by thomaseizinger** [comment]
Ugh, these two kademlia tests seems to be flaky as of recently. I'll deal with that tomorrow ..

**Comment by monoid** [comment]
Ok, I'm always at your disposal to merge/rebase/etc.

**Comment by thomaseizinger** [comment]
Okay, flaky tests are fixed on master, can you merge that into this branch please?

**Comment by monoid** [comment]
Done!

**Comment by mxinden** [comment]
Many CI steps are failing now that https://github.com/libp2p/rust-libp2p/pull/4620 is merged.

Given that `libp2p` now enables the `identity/rand` feature and given that we will release `libp2p` along with `libp2p-identity`, I am assuming that we can ignore the failures. Is that assumption correct @thomaseizinger?

**Comment by thomaseizinger** [comment]
> Many CI steps are failing now that #4620 is merged.

Yeah I realized that the situation is not ideal :sweat_smile: 
In other PRs, I've added internal changelog entries but that doesn't scale very well.

> I am assuming that we can ignore the failures.

You can't ignore them because they are marked as required. Mergify won't merge your PR until they pass. Let's brain-storm something later today.

**Review by thomaseizinger — APPROVED** [review]

**Comment by thomaseizinger** [comment]
@monoid There are no other PRs queued for merge at the moment so if you are around, we can get this one in soon :)

I'll need you to merge `master` in one more time. Unfortunately, GitHub doesn't allow me to update your branch because it is coming from an organisation. For future contributions (:pray:), things would be a bit easier if you'd have a personal fork :)

Thanks for this PR and sorry that it has been dragging on for so long!

**Comment by monoid** [comment]
Yeah, I know! Will make PR from the personal repo next time :)  I wasn't aware of this restriction (perhaps, it's just particular organization's security settings).

Thank you and other guys for the assistance and advices!

I've merged the master branch.  I will be online for the next 12h if any more action is needed on my side.

**Pull request merged**
